### PR TITLE
adding option for skipping SSL verification when using Git 

### PR DESCRIPTION
**Pull request by sethiyash**
#### What this PR does / why we need it:
#1286 

#### Which issue(s) this PR fixes:
#1286 

Fixes #

#### Does this PR introduce a user-facing change?

```release-note

```

##### Review Checklist:

- [ ] Follows the [developer guidelines](https://carvel.dev/shared/docs/latest/development_guidelines/)
- [ ] Relevant tests are added or updated
- [ ] Relevant docs in this repo added or updated
- [ ] Relevant carvel.dev docs added or updated in a separate PR and there's
  a link to that PR
- [ ] Code is at least as readable and maintainable as it was before this
  change

#### Additional documentation e.g., Proposal, usage docs, etc.:

```docs

```

### CODE REVIEW

**Review by praveenrewar — COMMENTED** [review]
The changes in the PR looks good to me. However, I think the idea was to use the existing `dangerousSkipTLSVerify` flag available in [kapp config](https://carvel.dev/kapp-controller/docs/latest/controller-config/#controller-configuration-spec), so we need to extend the functionality [here](https://github.com/carvel-dev/kapp-controller/blob/0594982604cb1a553bb12545cfa0d5a285617e5a/pkg/fetch/vendir.go#L382)

File: pkg/fetch/vendir.go [hunk 1]
```diff
@@ -421,3 +421,20 @@ func ExtractImageRegistry(name string) string {
...
+func extractGitHostname(input string) string {
+	u, err := url.Parse(input)
+	if err != nil {
+		return "github.com"
+	}
+	return u.Hostname()
+}
+
+// ExtractHost return registry for Docker Image and Host for git url
+func ExtractHost(input string, isGitURL bool) string {
```
Comment (praveenrewar) [inline pkg/fetch/vendir.go]: `isGirURL` doesn't sound very extensible, let's have a _type_ defined for the source type (like imgpkg or git) which we can pass down, and have switch case to extract the host, this way we can extend the functionality for other sources in future if needed.

**Review by praveenrewar — COMMENTED** [review]
Is it possible to add a test for this change?

File: pkg/fetch/vendir.go [hunk 2]
```diff
@@ -376,11 +386,9 @@ func (v *Vendir) configMapBytes(configMapRef vendirconf.DirectoryContentsLocalRe
 	return kyaml.Marshal(configMap)
 }
 
-// This function only works on image refs. If in the future we decide to
-// expand this option to other fetch options, we will need to add hostname
-// extraction for those
-func (v *Vendir) shouldSkipTLSVerify(url string) bool {
-	return v.opts.SkipTLSConfig.ShouldSkipTLSForAuthority(ExtractImageRegistry(url))
+// This function works on image refs and hostname extraction using isGitURL flag
```
Comment (praveenrewar) [inline pkg/fetch/vendir.go]: nit: Update comment

File: pkg/fetch/vendir_test.go [hunk 3]
```diff
@@ -58,32 +58,52 @@ func Test_AddDir_skipsTLS(t *testing.T) {
...
+		},
+		{
+			name:       "https://github.com/bitnami/charts/",
+			sourceType: fetch.GitURL,
+			want:       "github.com",
+		},
+		{
+			name:       "http://github.com/bitnami/charts/",
+			sourceType: fetch.GitURL,
+			want:       "github.com",
```
Comment (praveenrewar) [inline pkg/fetch/vendir_test.go]: Both test cases seem to be the same. Can we add a gitlab example as well?

File: pkg/fetch/vendir_test.go [hunk 4]
```diff
@@ -58,32 +58,52 @@ func Test_AddDir_skipsTLS(t *testing.T) {
 	}
 }
 
-func TestExtractImageRegistry(t *testing.T) {
```
Comment (praveenrewar) [inline pkg/fetch/vendir_test.go:61]: Let's also add a test similar to the one above (Test_AddDir_skipTLS) for git urls.

File: pkg/fetch/vendir.go [hunk 5]
```diff
@@ -21,7 +22,15 @@ import (
 	kyaml "sigs.k8s.io/yaml"
 )
 
+// SourceType to extract host
+type SourceType string
+
 const (
+	// GitURL source type to extract host
+	GitURL SourceType = "gitURL"
```
Comment (praveenrewar) [inline pkg/fetch/vendir.go]: I don't think we are going to use these string values, so we could probably just switch to int (Iota).

**Review by praveenrewar — COMMENTED** [review]
Adding an e2e test with a local gitlab server might be a bit much, but it would be great if we can specify what sort of manual testing has been done. (We follow [these](https://github.com/carvel-dev/vendir/pull/265#issuecomment-1746041098) steps and create an App which uses local gitlab server)

File: pkg/fetch/vendir.go [hunk 6]
```diff
@@ -22,6 +23,11 @@ import (
 )
 
 const (
+	// GitURL source type to extract host
+	GitURL = iota
+	// ImageRegistry source type to extract host
+	ImageRegistry
+
 	vendirEntireDirPath = "."
 )
```
Comment (praveenrewar) [inline pkg/fetch/vendir.go]: Should we keep these separate?

```suggestion
	ImageRegistry
)

const (
	vendirEntireDirPath = "."
)
```

File: pkg/fetch/vendir.go [hunk 7]
```diff
@@ -421,3 +428,24 @@ func ExtractImageRegistry(name string) string {
...
+func extractGitHostname(input string) string {
+	u, err := url.Parse(input)
+	if err != nil {
+		return "github.com"
+	}
+	return u.Hostname()
+}
+
+// ExtractHost return registry for Docker Image and Host for git url
+func ExtractHost(input string, sourceType int) string {
```
Comment (rohitagg2020) [inline pkg/fetch/vendir.go:442]: If this is internal function, we can rename it to start with smallcase.

File: pkg/fetch/vendir.go [hunk 8]
```diff
@@ -421,3 +428,24 @@ func ExtractImageRegistry(name string) string {
 	}
 	return registry
 }
+
+// extractGitHostname extracts the hostname from the git URL.
+func extractGitHostname(input string) string {
+	u, err := url.Parse(input)
+	if err != nil {
+		return "github.com"
```
Comment (rohitagg2020) [inline pkg/fetch/vendir.go]: If there is an error, why are we returning `github.com`?

File: pkg/fetch/vendir.go (hunk 7 above)
Comment (praveenrewar) [reply pkg/fetch/vendir.go:442]: It will be used in the test (which is a different package)

File: pkg/fetch/vendir.go (hunk 8 above)
Comment (sethiyash) [reply pkg/fetch/vendir.go]: I was thinking to default to github.com but we can return empty as well

File: pkg/fetch/vendir_test.go [hunk 9]
```diff
@@ -58,32 +58,95 @@ func Test_AddDir_skipsTLS(t *testing.T) {
...
+	}
+	for i, tc := range testCases {
+		err = vendir.AddDir(v1alpha1.AppFetch{
+			Git: &v1alpha1.AppFetchGit{URL: tc.URL},
+		},
+			"dirpath/0")
+		assert.NoError(t, err)
+
+		vConf := vendir.Config()
+		t.Logf("%+v", vConf)
```
Comment (praveenrewar) [inline pkg/fetch/vendir_test.go]: Remove log statements

**Review by praveenrewar — APPROVED** [review]
LGTM! (Apart from the minor comment)

**Pull request merged**